extensions/amp-analytics/0.1/test/test-linker.js

```diff
@@ -67,40 +79,74 @@ describe('Linker', () => {
         color: 'green',
         car: 'tesla',
       },
-      output: '1~1eit8v3~key1~value1~name~bob~color~green~car~tesla',
+      output: '1*efercj*key1*dmFsdWUx*name*Ym9i*color*Z3JlZW4.*car*dGVzbGE.',
     },
     {
-      description: 'fake macros',
+      description: 'encodes URL unsafe chars in values',
       version: '1',
       pairs: {
         cid: '12345',
         ref: 'https://www.example.com',
       },
-      output: '1~s21iu~cid~12345~ref~https%3A%2F%2Fwww.example.com',
+      output: '1*vz1vnu*cid*MTIzNDU.*ref*aHR0cHM6Ly93d3cuZXhhbXBsZS5jb20.',
+    },
+    {
+      description: 'encodes * in values',
+      version: '1',
+      pairs: {
+        'key': '*hi*',
+        'key2': '***',
+      },
+      output: '1*lpias9*key*KmhpKg..*key2*Kioq',
+    },
+    {
+      description: 'encodes unicode in values',
+      version: '1',
+      pairs: {
+        'key': '中文',
+        'key2': 'π',
+      },
+      output: '1*1pf8zsw*key*5Lit5paH*key2*z4A.',
     },
     {
-      description: 'encodes url safe keys and values',
+      description: 'allows base64 chars in keys',
       version: '1',
       pairs: {
-        '<unsafe>': '//foo@bar',
+        '0x3': '0x3',
+        '_gb': '_gb',
+        'g.b': 'g.b',
+        'nn-': 'nn-',
       },
-      output: '1~1gbjnst~%3Cunsafe%3E~%2F%2Ffoo%40bar',
+      output: '1*10d75fz*0x3*MHgz*_gb*X2di*g.b*Zy5i*nn-*bm4t',
     },
     {
-      description: 'encodes tilde in key value pairs',
+      description: 'ignore invalid keys',
       version: '1',
       pairs: {
-        '~key~': 'hi',
-        'key2': '~hi~',
+        '*invalid': '123',
+        'valid': 'abc',
       },
-      output: '1~1wqggkq~%7Ekey%7E~hi~key2~%7Ehi%7E',
+      output: '1*bn7mun*valid*YWJj',
+      expectErrors: 1,
+    },
+    {
+      description: 'return empty string if all keys are invalid',
+      version: '1',
+      pairs: {
+        '*invalid': '123',
+        'invalid!': 'abc',
+      },
+      output: '',
+      expectErrors: 2,
     },
   ];
 
   tests.forEach(test => {
     it(test.description, () => {
-      const result = createLinker(test.version, test.pairs);
-      return expect(result).to.equal(test.output);
+      if (test.expectErrors) {
+        expectAsyncConsoleError(/Invalid linker key/, test.expectErrors);
```

Comment:
Why do we need async here? Seems like everything should be sync?

Reply:
`expectAsyncConsoleError` is easier to do conditional here comparing to allowConsoleErrors